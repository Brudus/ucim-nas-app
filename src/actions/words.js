import database from '../firebase/firebase';

// ADD_WORD
export const addWord = (word) => ({
    type: 'ADD_WORD',
    word
});

export const startAddWord = (wordData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            source = '', 
            destination= '', 
            repeatAt = 0,
            interval = 0,
            reps = 0,
            easeFactor = 2.5,
            repeatAtInverted = 0,
            intervalInverted = 0,
            repsInverted = 0,
            easeFactorInverted = 2.5
        } = wordData;
        const word = { source, destination, repeatAt, interval, reps, easeFactor, repeatAtInverted, intervalInverted, repsInverted, easeFactorInverted };
        
        return database.ref(`users/${uid}/words`).push(word).then((ref) => {
            dispatch(addWord({
                id: ref.key,
                ...word
            }));
        });
    };
};

// REMOVE_WORD
export const removeWord = ({ id } = {}) => ({
    type: 'REMOVE_WORD',
    id
});

export const startRemoveWord = ({ id } = {}) => {
    return (dispatch,getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/words/${id}`).remove().then(() => {
            dispatch(removeWord({ id }));
        });
    };
};

// EDIT_WORD
export const editWord = (id, updates) => ({
    type: 'EDIT_WORD',
    id,
    updates
});

export const startEditWord = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/words/${id}`).update(updates).then(() => {
            dispatch(editWord(id, updates));
        })
    };
};

// SET_WORDS
export const setWords = (words) => ({
    type: 'SET_WORDS',
    words
});

export const startSetWords = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/words`).once('value').then((snapshot) => {
            const words = [];

            snapshot.forEach((childSnapshot) => {
                words.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            
            dispatch(setWords(words));
        });
    };
};