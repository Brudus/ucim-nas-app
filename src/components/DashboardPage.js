import React from 'react';

const DashboardPage = () => (
    <div>
        <p>You are logged in!</p>
    </div>
);

export default DashboardPage;