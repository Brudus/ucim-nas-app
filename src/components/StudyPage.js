import React from 'react';
import { connect } from 'react-redux';

export const StudyPage = (props) => (
    <div>
        <h2>Study!</h2>
        {props.words[0].source}
    </div>
);

const mapStateToProps = (state) => ({
    words: state.words
});

export default connect(mapStateToProps)(StudyPage);