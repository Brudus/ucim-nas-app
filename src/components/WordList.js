import React from 'react';
import { connect } from 'react-redux';
import WordListItem from './WordListItem';
import selectWords from '../selectors/words';

export const WordList = (props) => (
    <div>
        {
            props.words.length === 0 ? (
                <p>No words available.</p>
            ) : (
                props.words.map((word) => <WordListItem key={word.id} {...word} />)
            )
        }
        { }
    </div>
);

const mapStateToProps = (state) => {
    return {
        words: selectWords(state.words, state.filters)
    };
};

export default connect(mapStateToProps)(WordList);