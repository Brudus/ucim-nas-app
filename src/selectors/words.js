export default (words, { text, sortBy }) => {
    return words.filter((word) => {
        const sourceText = word.source.toLowerCase()
        const destinationText = word.destination.toLowerCase();
        const searchText = text.toLowerCase().trim();

        const sourceMatch = sourceText.includes(searchText);
        const destinationMatch = destinationText.includes(searchText);
        return sourceMatch || destinationMatch;
    }).sort((a, b) => {
        if (sortBy === 'repeatAt') {
            return a.repeatAt > b.repeatAt ? 1 : -1;
        } else if (sortBy === 'source') {
            const aText = a.source.toLowerCase().trim();
            const bText = b.source.toLowerCase().trim();
            return aText > bText ? 1 : -1;
        } else if (sortBy === 'destination') {
            const aText = a.destination.toLowerCase().trim();
            const bText = b.destination.toLowerCase().trim();
            return aText > bText ? 1 : -1;
        }
    });
};