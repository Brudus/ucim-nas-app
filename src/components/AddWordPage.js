import React from 'react';
import { connect } from 'react-redux';
import WordForm from './WordForm';
import { startAddWord } from '../actions/words';

export class AddWordPage extends React.Component {
    onSubmit = (word) => {
        this.props.startAddWord(word);
    };

    render() {
        return (
            <div>
                <h1>Add Word</h1>
                <WordForm 
                    onSubmit={this.onSubmit}
                />
            </div>
        );
    };
}

const mapDispatchToProps = (dispatch) => ({
    startAddWord: (word) => dispatch(startAddWord(word))
});

export default connect(undefined, mapDispatchToProps)(AddWordPage);