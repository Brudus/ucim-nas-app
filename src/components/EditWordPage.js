import React from 'react';
import { connect } from 'react-redux';
import WordForm from './WordForm';
import { startEditWord, startRemoveWord } from '../actions/words';

export class EditWordPage extends React.Component {
    onSubmit = (word) => {
        this.props.startEditWord(this.props.word.id, word);
        this.props.history.push('/catalog');
    };
    onRemove = (id) => {
        this.props.startRemoveWord({ id: this.props.word.id });
        this.props.history.push('/catalog');
    };

    render() {
        return (
            <div>
                <WordForm 
                    word={this.props.word}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onRemove}>Remove</button>
            </div>
        );
    };
};

const mapStateToProps = (state, props) => ({
    word: state.words.find((word) => word.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
    startEditWord: (id, word) => dispatch(startEditWord(id, word)),
    startRemoveWord: (data) => dispatch(startRemoveWord(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditWordPage);