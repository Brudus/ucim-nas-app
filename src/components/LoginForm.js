import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export class LoginForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            email: '',
            password: '',
            loginError: '',
            emailPassed: undefined
        };
    }

    validateEmail = (email) => {
        const regex = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
        return regex.test(email);
    };

    onEmailChange = (e) => {
        const email = e.target.value;
        const emailPassed = this.validateEmail(email);
        this.setState(() => ({ 
            email,
            emailPassed
        }));
    };

    onPasswordChange = (e) => {
        const password = e.target.value;
        this.setState(() => ({ password }));
    }

    onSubmit = (e) => {
        e.preventDefault();

        this.setState(() => ({ 
            loginError: '',
            emailPassed: true
        }));
        this.props.startLogin(this.state.email, this.state.password);
    };

    render() {
        return (
            <div>
                <form onSubmit={this.onSubmit}>
                    {this.state.loginError && <p>{this.state.loginError}</p>}
                    <div>
                        <input 
                            type="email"
                            placeholder="Email address" 
                            autoFocus
                            onChange={this.onEmailChange}
                        />
                    </div>
                    
                    <div>
                        <input 
                            type="password"
                            placeholder="Password" 
                            onChange={this.onPasswordChange}
                        />
                    </div>
                    <button disabled={!this.state.emailPassed || !this.state.password}>Login</button>
                </form> 
            </div>
        );
    };
};

const mapDispatchToProps = (dispatch, props) => ({
    startLogin: (email, password) => dispatch(startLogin(email, password))
});

export default connect(undefined, mapDispatchToProps)(LoginForm);