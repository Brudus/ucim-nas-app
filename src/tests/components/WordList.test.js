import React from 'react';
import { shallow } from 'enzyme';
import { WordList } from '../../components/WordList';
import words from '../fixtures/words';

test('should render WordList with words', () => {
    const wrapper = shallow(<WordList words={words} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render WordList with empty message', () => {
    const wrapper = shallow(<WordList words={[]} />);
    expect(wrapper).toMatchSnapshot();
});