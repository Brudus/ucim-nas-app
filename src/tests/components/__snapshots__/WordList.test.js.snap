// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render WordList with empty message 1`] = `
<div>
  <p>
    No words available.
  </p>
</div>
`;

exports[`should render WordList with words 1`] = `
<div>
  <WordListItem
    destination="dobro"
    factor={1}
    id="1"
    key="1"
    repeatAt={0}
    source="gut"
  />
  <WordListItem
    destination="nisam"
    factor={1.5}
    id="2"
    key="2"
    repeatAt={1024}
    source="bin ich nicht"
  />
  <WordListItem
    destination="hvala"
    factor={3}
    id="3"
    key="3"
    repeatAt={-1000}
    source="danke"
  />
</div>
`;
